File: transcribers/assemblyai.js
/**
 * AssemblyAI transcription service for Vercel serverless
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');

class AssemblyAITranscriber {
  constructor(apiKey) {
    if (!apiKey) {
      throw new Error('AssemblyAI API key is required');
    }
    this.apiKey = apiKey;
    this.baseUrl = 'https://api.assemblyai.com/v2';
    this.name = 'AssemblyAI';
    this.language = 'lv';
  }

  /**
   * Upload file to AssemblyAI
   */
  async uploadFile(fileBuffer) {
    return new Promise((resolve, reject) => {
      const options = {
        method: 'POST',
        hostname: 'api.assemblyai.com',
        path: '/v2/upload',
        headers: {
          'authorization': this.apiKey,
          'content-type': 'application/octet-stream',
          'content-length': fileBuffer.length
        }
      };

      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          if (res.statusCode === 200) {
            const result = JSON.parse(data);
            resolve(result.upload_url);
          } else {
            reject(new Error(`Upload failed: ${res.statusCode} - ${data}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(new Error(`Upload request failed: ${error.message}`));
      });

      req.write(fileBuffer);
      req.end();
    });
  }

  /**
   * Request transcription
   */
  async requestTranscription(audioUrl) {
    return new Promise((resolve, reject) => {
      const postData = JSON.stringify({
        audio_url: audioUrl,
        language_code: 'lv',
        speech_model: 'best'
      });

      const options = {
        method: 'POST',
        hostname: 'api.assemblyai.com',
        path: '/v2/transcript',
        headers: {
          'authorization': this.apiKey,
          'content-type': 'application/json',
          'content-length': Buffer.byteLength(postData)
        }
      };

      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          if (res.statusCode === 200) {
            const result = JSON.parse(data);
            resolve(result.id);
          } else {
            reject(new Error(`Transcription request failed: ${res.statusCode} - ${data}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(new Error(`Transcription request failed: ${error.message}`));
      });

      req.write(postData);
      req.end();
    });
  }

  /**
   * Poll for transcription completion
   */
  async waitForCompletion(transcriptId, maxWait = 300000) {
    const startTime = Date.now();
    const pollInterval = 3000; // 3 seconds

    while (Date.now() - startTime < maxWait) {
      const result = await this.checkStatus(transcriptId);
      
      if (result.status === 'completed') {
        if (!result.text) {
          throw new Error('No transcription text returned');
        }
        return result.text;
      } else if (result.status === 'error') {
        throw new Error(`Transcription failed: ${result.error || 'Unknown error'}`);
      }

      // Still processing - wait before polling again
      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Transcription timed out after ${maxWait / 1000} seconds`);
  }

  /**
   * Check transcription status
   */
  async checkStatus(transcriptId) {
    return new Promise((resolve, reject) => {
      const options = {
        method: 'GET',
        hostname: 'api.assemblyai.com',
        path: `/v2/transcript/${transcriptId}`,
        headers: {
          'authorization': this.apiKey
        }
      };

      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          if (res.statusCode === 200) {
            resolve(JSON.parse(data));
          } else {
            reject(new Error(`Status check failed: ${res.statusCode} - ${data}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(new Error(`Status check request failed: ${error.message}`));
      });

      req.end();
    });
  }

  /**
   * Main transcribe method
   */
  async transcribe(fileBuffer) {
    try {
      console.log('Starting AssemblyAI transcription...');
      
      // Step 1: Upload file
      const uploadUrl = await this.uploadFile(fileBuffer);
      console.log('File uploaded successfully');
      
      // Step 2: Request transcription
      const transcriptId = await this.requestTranscription(uploadUrl);
      console.log(`Transcription requested, ID: ${transcriptId}`);
      
      // Step 3: Wait for completion
      const transcript = await this.waitForCompletion(transcriptId);
      console.log('Transcription completed successfully');
      
      return transcript;
    } catch (error) {
      console.error('AssemblyAI transcription error:', error);
      throw new Error(`AssemblyAI transcription failed: ${error.message}`);
    }
  }
}

module.exports = AssemblyAITranscriber;

File: transcribers/assemblyai.py
"""AssemblyAI transcription service."""

import os
import time
import requests
from pathlib import Path
from typing import Optional

from .base import BaseTranscriber, TranscriptionError


class AssemblyAITranscriber(BaseTranscriber):
    """AssemblyAI transcription service with Latvian support."""
    
    def __init__(self, api_key: Optional[str] = None):
        super().__init__(
            name="AssemblyAI",
            language="lv"
        )
        
        self.api_key = api_key or os.getenv('ASSEMBLYAI_API_KEY')
        if not self.api_key:
            raise ValueError("AssemblyAI API key is required")
        
        self.base_url = "https://api.assemblyai.com/v2"
        self.headers = {
            "authorization": self.api_key,
            "content-type": "application/json"
        }
    
    def transcribe(self, audio_file_path: Path) -> str:
        """Transcribe audio using AssemblyAI."""
        if not self.validate_audio_file(audio_file_path):
            raise TranscriptionError(f"Invalid audio file: {audio_file_path}")
        
        try:
            self.logger.info(f"Starting AssemblyAI transcription for: {audio_file_path}")
            
            # Step 1: Upload the audio file
            upload_url = self._upload_file(audio_file_path)
            self.logger.info(f"File uploaded successfully")
            
            # Step 2: Request transcription
            transcript_id = self._request_transcription(upload_url)
            self.logger.info(f"Transcription requested, ID: {transcript_id}")
            
            # Step 3: Poll for completion
            transcript = self._wait_for_completion(transcript_id)
            self.logger.info(f"Transcription completed successfully")
            
            return transcript
            
        except Exception as e:
            self.logger.error(f"AssemblyAI transcription failed: {str(e)}")
            raise TranscriptionError(f"AssemblyAI transcription failed: {str(e)}")
    
    def _upload_file(self, file_path: Path) -> str:
        """Upload audio file to AssemblyAI and get upload URL."""
        upload_endpoint = f"{self.base_url}/upload"
        
        with open(file_path, 'rb') as f:
            response = requests.post(
                upload_endpoint,
                headers={"authorization": self.api_key},
                data=f
            )
        
        if response.status_code != 200:
            raise TranscriptionError(f"Upload failed: {response.status_code} - {response.text}")
        
        return response.json()['upload_url']
    
    def _request_transcription(self, audio_url: str) -> str:
        """Request transcription from AssemblyAI."""
        transcript_endpoint = f"{self.base_url}/transcript"
        
        json_data = {
            "audio_url": audio_url,
            "language_code": "lv",  # Latvian language
            "speech_model": "best"  # Use the best available model
        }
        
        response = requests.post(
            transcript_endpoint,
            json=json_data,
            headers=self.headers
        )
        
        if response.status_code != 200:
            raise TranscriptionError(f"Transcription request failed: {response.status_code} - {response.text}")
        
        return response.json()['id']
    
    def _wait_for_completion(self, transcript_id: str, max_wait: int = 300) -> str:
        """Poll AssemblyAI API until transcription is complete."""
        polling_endpoint = f"{self.base_url}/transcript/{transcript_id}"
        
        start_time = time.time()
        
        while time.time() - start_time < max_wait:
            response = requests.get(polling_endpoint, headers=self.headers)
            
            if response.status_code != 200:
                raise TranscriptionError(f"Polling failed: {response.status_code} - {response.text}")
            
            result = response.json()
            status = result['status']
            
            if status == 'completed':
                # Ensure UTF-8 encoding for Latvian characters
                transcript = result.get('text', '')
                if not transcript:
                    raise TranscriptionError("No transcription text returned")
                return transcript
            
            elif status == 'error':
                error_msg = result.get('error', 'Unknown error')
                raise TranscriptionError(f"Transcription failed: {error_msg}")
            
            # Still processing - wait before polling again
            time.sleep(3)
        
        raise TranscriptionError(f"Transcription timed out after {max_wait} seconds")

